### ISSUE TITLE
Interference with kubectl auth plugin (kubelogin)

### ISSUE THREAD

**Issue by bvergnaud** [kubecolor/kubecolor#65]
Hello 👋🏻 

In my workplace, we use dex and [the kubelogin plugin](https://github.com/int128/kubelogin) to provide authentification on kube's API.
Unfortunately, kubecolor does not handle that very gracefully. When not logged in, kubelogin will show prompts for your username and password, both of which are captured by kubecolor, giving the illusion that the command hangs forever.

Once you know that, you can blindly input your username and password, and it works fine. It's just not very intuitive.

Obviously, I'm describing my use case, but I expect similar problems would occur with any plugin that produces non standard output with some level of interactivity.

# Expected Behavior

Ideally, kubecolor should find a way to let non-standard output through, or maybe to flush the output regularly instead of holding everything back until the full output is produced by kubectl. I don't really mind the output of plugins being colorized or not.

I'm not familiar with kubecolor's codebase at all but I understand that it might not be trivial or even possible at all for kubecolor to separate native / non-native output, so I'm not sure what the best approach would be to fix this issue.

# Current Behavior

Kubecolor captures everything and does not seem to flush anything until it has the full output.

![image](https://github.com/kubecolor/kubecolor/assets/9599845/db988283-5813-4a38-87e4-b06d45d350d4)
In this screenshot, "bvergnaud" is me typing my username in the void (as well as my password which is still hidden thankfully), and "Username: Password: " are the authentication prompts captured and colorized, being printed after I've logged in, along the normal/expected output of my kubectl command.

## Context

kubecolor 0.2.2
kubectl 1.29.1
kubelogin 1.28.0

**Comment by applejag** [comment]
Hello @bvergnaud, thanks for reporting this! We've not considered this edge case yet.

Kubecolor does detect when a sub-command is a kubectl plugin (such as from krew), but what's happening here is that kubectl is using a credentials plugin, which is harder to detect.

Then most commands, such as the `kubectl get` sub-command, uses a bufio.Scanner, which patiently waits for a newline character. The input fields here are printed without newline characters, which results in them never being printed.

We also use Dex and int128's kubelogin at my workplace, but we use OIDC, so we get redirected to a browser instead. I've missed this use case where you enter credentials directly to kubectl/kubelogin. Definitely a bug in kubecolor.

Spontaneously I don't have a fix in mind. So you'll sadly have to continue using your workaround until we can think of a fix.

**Comment by bvergnaud** [comment]
Thanks for the feedback and explanations !
And no worries, I'm in no rush to get this fixed. I'll just be watching this issue patiently. 😄

**Comment by larsgerber** [comment]
Hello @applejag

Thanks for maintaining this cool project. I have a question about Krew and the statement that subcommands are not colored. In https://github.com/kubecolor/kubecolor/pull/8 this feature has been added. As far as I understand it, it looks for the path prefix kubectl- to determine if its a subcommand. But this does not work for me, as can be seen in the screenshot below. Is this a bug or is it only possible if the plugin is started directly as a script?

Example
![Screenshot 2024-04-22 at 15 44 55](https://github.com/kubecolor/kubecolor/assets/75072836/3d5b88ab-5926-4820-a914-753bb38d9c8e)

In my case, fzf is not able to show a preview of it.
```
DEBUG=true ./kubectl-node_connect
+ set -eou pipefail
+ IFS='
	'
+ kubecolor get pods
+ fzf
```

kubecolor 0.3.0 (Brew)
kubectl v1.30.0 (Brew)
fzf 0.50.0 (Brew)

**Comment by applejag** [comment]
Thanks for reporting that, @larsgerber! However, I created a new issue for this, as the issue you're having is unrelated to this issue. Let's move the discussion there.

**Comment by crabique** [comment]
Not sure if related, but there is inconsistent behaviour in terms of plugins output renderer not respecting the TTY:
```bash
❯ kubectl get secret test --template '{{ base64decode .data.pass }}' | bat -A # good
  testpassword
  
❯ kubectl view-secret test pass | bat -A # bad, not usable in $() because of the escape seqs
  ␛[32mtestpassword␛[0m␊
  
❯ \kubectl view-secret test pass | bat -A # good
  testpassword
```

**Comment by applejag** [comment]
> Not sure if related, but there is inconsistent behaviour in terms of plugins output renderer not respecting the TTY:
> 
> ```shell
> ❯ kubectl get secret test --template '{{ base64decode .data.pass }}' | bat -A # good
>   testpassword
>   
> ❯ kubectl view-secret test pass | bat -A # bad, not usable in $() because of the escape seqs
>   ␛[32mtestpassword␛[0m␊
>   
> ❯ \kubectl view-secret test pass | bat -A # good
>   testpassword
> ```

That's unrelated. But still interesting. Please create a new issue for this one so we can look into it further

**Comment by applejag** [comment]
Did some research an testing around with some solutions, and the way forward seems to be to make use of non-blocking IO.

For context: When reading there's no syscall that says "read until the next newline", so we are forced to buffer the output. And when buffering the output we are basically guaranteed to get buffered output ending in the middle of a line. Without non-blocking IO if we read then kubecolor will wait until kubectl prints the rest of the line, causing the "nothing is printed" problem that this issue is about.

But with non-blocking IO we can do a "try read", and if the OS says "hey there's nothing more to read yet" then we know we should try to print what we have for now.

We can't always just "print what we have for now" as a lot of our parsing relies on later text in the output. For example if a line from `kubectl get pod -o yaml` says `foo: n`, then if we know the line ends there then we print the value (`n`) with "string color", but if the line continues with `ull` to give `foo: null` then we need to print the value (`null`) with "null color".

---

Alright, so from my testing I found the following approach using non-blocking IO will work decently well:

1. set the pipe to non-blocking using [syscall.SetNonblock](https://pkg.go.dev/syscall#SetNonblock). Pipes from Go seem to created using non-blocking IO, so needs to be verified if this is really needed.
2. read using normal `io.Reader.Read()`
3. if the data we read does not contain `\n`, then we want to see if it's an unfinished line printed:   1. wait 1ms. In my testing sometimes two writes from the kubectl auth exec program can take some time (~10-20us (not ms), such as by two separate `echo`: `echo -n "hello" && echo -n "world"`) so we want to make sure we got the entirety of the pending line (aside: I wonder if this delay might be bigger on `kubectl logs` output as the data is being streamed from the server... Needs further testing! Possibly we could resolve it by using different timeout for different use cases, such as 10ms on kubectl stderr logs, but 200ms on stdout)
   2. do an unblocking read using [syscall.Read](https://pkg.go.dev/syscall#Read)
   3. if that read returns error `EAGAIN` then we know that's all that was printed for the line so far, and so we should render it in kubecolor.
4. repeat from step 2.

This was sadly only working on Linux. I wrote a small POC code which did work well on Linux. However, on Windows, the [`syscall.SetNonblock` is not implemented](https://cs.opensource.google/go/go/+/refs/tags/go1.24.1:src/syscall/exec_windows.go;l=149-151;drc=5351bcf8225747f0ef39afc44c0499822992ed11), and when I tried my POC code on my Windows laptop it sure did not work. I have no idea how this behaves on Mac. It would probably be wise to have a fallback implementation that depends on channels or something like that, similar to the existing [kubecolor `ctxscanner`](https://github.com/kubecolor/kubecolor/blob/a0574acc1489b1b68e78b82fcaf06d0aee078f98/internal/ctxscanner/ctxscanner.go), just to make sure things also work for other platforms (e.g BSD or WASM). And then have more optimized implementations when possible using non-blocking IO.

The pgx project implemented non-blocking IO on Windows in https://github.com/jackc/pgx/pull/1525, which can be seen in <https://github.com/jackc/pgx/blob/2f8cfe9950c2937ab122eb07ba33ef0edf3f5166/internal/nbconn/nbconn_real_non_block_windows.go>. They're MIT licensed so could try take inspiration from that.

This whole approach requires a bunch of refactoring because our parsers like `logscan` and `describescan` use `bufio.Scanner` and therefore are written to assume they read one line at a time. They need to be changed so they support reading only parts of a line first so it can print parts of it, and then the rest of the same line later.

While implementing this I'll need to also see how #184 and #201 are affected by this, as replacing the `bufio.Scanner` will probably also resolve both of those as well. On cut lines like this or super long lines (such as what's causing the bug in #184) we might also just print them as-is without any coloring.

**Comment by applejag** [comment]
Go 1.25 will support async IO on Windows: https://tip.golang.org/doc/go1.25#ospkgos

The implementation should still have a channel-based (or similar) fallback, but we should be able to lean on async IO even for Windows

**Comment by mroelofsen-acc** [comment]
**Reproduced with Pinniped + diagnostic proof of fd type**

We hit this same issue with [Pinniped](https://pinniped.dev)'s `cli_password` flow and ran a diagnostic that pinpoints the exact mechanism.

**Setup:** a minimal exec credential plugin that writes a string *without* a newline to stderr, sleeps 3 seconds, then writes a string *with* a newline. Run through both `/usr/bin/kubectl` and `/usr/bin/kubecolor` using pexpect, with the plugin logging its own fd types via `os.fstat`.

**Results:**

| | `/usr/bin/kubectl` | `/usr/bin/kubecolor` |
|---|---|---|
| stderr fd type seen by plugin | `char-device (tty=True)` | `pipe` |
| Prompt appears before 3s sleep | yes — immediately | no — only after the newline arrives |

Real kubectl passes stderr to the exec plugin as the inherited TTY fd. Writes appear immediately regardless of newlines.

Kubecolor pipes stderr from kubectl. That pipe becomes the exec plugin's stderr fd. Writes without trailing newlines are held in the pipe buffer until a newline arrives — which in Pinniped's case only happens after both credential reads complete, so both `Username: ` and `Password: ` appear together after the user has already typed blind.

This explains why `"Log in to activedirectory\n"` appears correctly (ends with `\n`) while the prompts don't.

Tested on kubecolor v0.5.3 and confirmed not fixed in v0.6.0.

---

**PR #280** (`bugfix/fix-line-reading`) is directly working on this fix — see our comment there for a concrete `runner.go` implementation suggestion using `bufio.Reader.Buffered() == 0` to flush partial lines immediately while preserving colorization for complete lines: https://github.com/kubecolor/kubecolor/pull/280#issuecomment-4407016097
